### PR TITLE
Draft: Reduce template bloat in std.format

### PR DESCRIPTION
**Pull request by nordlow**
This, for instance, reduces memory usage of unittest-checking `std.format.write` as

```sh
dmd -o- -vcolumns -wi -debug -unittest -dip25 -dip1000 -version\=StdUnittest -I/home/per/Work/phobos/ /home/per/Work/phobos/std/format/write.d
```

from 386 MB to 356 MB.

1. Motive of added const qualifers is to have, for instance,

```d
format(fmt, int.init)
format(fmt, const(int).init)
format(fmt, immutable(int).init)
...
```

all result in the same template instance of `format`.

2. I've factored out integer printing to make the code not be instantiated for each integral type potentially reducing generated code a lot when formatting and printing integrals. Separate PR at https://github.com/dlang/phobos/pull/8278.

3. The refactoring in `formatValue` avoids the need to instantiate the wrapper templates `BooleanTypeOf`, `IntegralTypeOf`, etc.

I'm gonna look into other types classes such as floating point types next.

Putting this out into the open to see what CIs feel about this.

Feel free to tell me how you want this split up.

### CODE REVIEW

File: std/format/write.d [hunk 1]
```diff
@@ -1228,15 +1228,25 @@ Note:
 See_Also:
     $(LREF formattedWrite) which formats several values at once.
  */
-void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, scope const ref FormatSpec!Char f)
+void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, scope const ref FormatSpec!Char f) // TODO: make const(T) work
```
Comment (nordlow) [inline std/format/write.d:1231]: Should we qualify `T` as `const` here? If I do I get a strange breakage in another Phobos module.

**Comment by nordlow** [comment]
`make -f posix.mak unittest` passes locally.

File: std/format/internal/write.d [hunk 2]
```diff
@@ -28,11 +28,15 @@ package(std.format):
     `bool`s are formatted as `"true"` or `"false"` with `%s` and as `1` or
     `0` with integral-specific format specs.
  */
-void formatValueImpl(Writer, T, Char)(auto ref Writer w, T obj, scope const ref FormatSpec!Char f)
+void formatValueImpl(Writer, T, Char)(auto ref Writer w, const(T) obj, scope const ref FormatSpec!Char f) // const(T) reduces number of instances
```
Comment (nordlow) [inline std/format/internal/write.d:31]: Ok to remove the comment if needed.

File: std/format/internal/write.d [hunk 3]
```diff
@@ -28,15 +28,19 @@ package(std.format):
...
+    formatValueImplBool!(Writer, Char)(w, val, f);
+}
 
+void formatValueImplBool(Writer, Char)(auto ref Writer w, const(bool) val, scope const ref FormatSpec!Char f)
+{
     if (f.spec == 's')
         writeAligned(w, val ? "true" : "false", f);
     else
-        formatValueImpl(w, cast(byte) val, f);
+        formatValueImplIntegral(w, cast(const(byte))val, f);
```
Comment (RazvanN7) [inline std/format/internal/write.d:43]: The style checker complains about such casts. A space should be between the cast and the value that is being cast `cast(type) val`.

**Comment by RazvanN7** [comment]
Irrespective of the memory gains, I think that having the argument to format being `const` makes sense on its own as I cannot imagine any scenario where the argument might get modified.

File: std/format/internal/write.d (hunk 3 above)
Comment (nordlow) [reply std/format/internal/write.d]: This also applies to code in comments...

File: std/format/internal/write.d (hunk 2 above)
Comment (RazvanN7) [reply std/format/internal/write.d:31]: I think we should remove them. As I mentioned before, having the argument be `const` makes sense from an interface point of view. `format` should not modify the value that is being inserted.

File: std/format/internal/write.d [hunk 4]
```diff
@@ -157,21 +161,25 @@ if (is(immutable T == immutable typeof(null)) && !is(T == enum) && !hasToString!
...
  */
-void formatValueImpl(Writer, T, Char)(auto ref Writer w, T obj, scope const ref FormatSpec!Char f)
+void formatValueImpl(Writer, T, Char)(auto ref Writer w, const(T) obj, scope const ref FormatSpec!Char f) // const(T) reduces number of instances
 if (is(IntegralTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
 {
-    import std.range.primitives : put;
-
     alias U = IntegralTypeOf!T;
-    U val = obj;    // Extracting alias this may be impure/system/may-throw
+    U val = obj;         // Extracting alias this may be impure/system/may-throw
```
Comment (RazvanN7) [inline std/format/internal/write.d:168]: This change is unrelated to this PR

File: std/format/internal/write.d [hunk 5]
```diff
@@ -28,15 +28,19 @@ package(std.format):
...
-void formatValueImpl(Writer, T, Char)(auto ref Writer w, T obj, scope const ref FormatSpec!Char f)
+void formatValueImpl(Writer, T, Char)(auto ref Writer w, const(T) obj, scope const ref FormatSpec!Char f) // const(T) reduces number of instances
 if (is(BooleanTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
 {
-    BooleanTypeOf!T val = obj;
+    BooleanTypeOf!T val = obj;  // TODO: remove and pass obj to val directly in line below?
+    formatValueImplBool!(Writer, Char)(w, val, f);
+}
 
+void formatValueImplBool(Writer, Char)(auto ref Writer w, const(bool) val, scope const ref FormatSpec!Char f)
```
Comment (RazvanN7) [inline std/format/internal/write.d:38]: I fail to understand how factoring code out into a new template declaration reduces template bloat

Comment (RazvanN7) [reply std/format/internal/write.d:38]: Oh, I understand now. Code that is the same for each template instance should be factored out so that the generated code is smaller.

Comment (RazvanN7) [reply std/format/internal/write.d:38]: Honestly, I think that such optimizations should be benchmarked on their own. It is hard to assess the impact of this optimization when it is bundled with others. I suggest splitting this PR into different ones and assess the gains on their own. There must be a proven benefit to accept such code surgery.

File: std/format/internal/write.d (hunk 4 above)
Comment (nordlow) [reply std/format/internal/write.d]: Done.

File: std/format/internal/write.d (hunk 3 above)
Comment (nordlow) [reply std/format/internal/write.d]: Fixed.

File: std/format/internal/write.d (hunk 2 above)
Comment (nordlow) [reply std/format/internal/write.d:31]: Ok, I'll remove them.

File: std/format/internal/write.d (hunk 5 above)
Comment (nordlow) [reply std/format/internal/write.d:38]: Ok.

File: std/format/write.d (hunk 1 above)
Comment (nordlow) [reply std/format/write.d]: Skipping this.

**Comment by nordlow** [comment]
So far this PR reduces number of instances of formatValue from 10890 to 5459.

**Comment by nordlow** [comment]
> @nordlow this is currently failing the testing pipeline.

Thanks. I'm gonna branch of minor changes from this. So I'm gonna wait with fixing this.

**Comment by nordlow** [comment]
The two issues with const-qualifying the parameters of `format`, `text`, etc are that

1. 

```d
auto format(T...)(const(T) args) {}
```

fails when any of the `args` is a struct with at least one field being a `const class` (which is related to the lack of expressing head-mutable class fields in D without resorting to the `Rebindable`-hack which, btw, contamintes logic in `hasAliasing`). There is currently no test for this case in Phobos so those need to be added aswell.

2. 

```d
auto format(T...)(const T args) {}
```

fails when any of the `args` is an `InputRange`. The same holds for `std.format.text()` and `traceXYZ` functions in `std.experimental.logging`.

Both are related to lack of implicit conversion from const to tail-const. @andralex has previously eluded to that those rules need relaxing.

I'm personally not sure which alternative is the best long-term.

FYI, @adamdruppe.

**Comment by RazvanN7** [comment]
@nordlow I don't know of any plans to introduce tail-constness into the language, so I suggest, in the interest of progress to implement in this PR whatever works and is not dependent on tail-constness. Otherwise, this will be stalled forever.

**Comment by LightBender** [comment]
PR closed as stalled. This can be resurrected for Phobos 3.
